core/modules/replica/SqlRequest.h

```diff
@@ -46,57 +48,173 @@ namespace qserv {
+     * @return a reference to a result obtained from a remote service.
      *
      * @note
      *   This operation will return a sensible result only if the operation
      *   finishes with status FINISHED::SUCCESS
      */
+    SqlResultSet const& responseData() const;
+
+    /// @see Request::extendedPersistentState()
+    std::list<std::pair<std::string,std::string>> extendedPersistentState() const final;
+
+protected:
+
     /**
-     * @return
-     *   request-specific extended data reported upon a successful
-     *   completion of the request
+     * Create a new request with specified parameters.
+     *
+     * @param serviceProvider
+     *   is needed to access the Configuration and the Controller for communicating
+     *   with the worker
+     *
+     * @param io_service
+     *   a communication end-point
+     *
+     * @param worker
+     *   identifier of a worker node
+
+     * @param maxRows
+     *   (optional) limit for the maximum number of rows to be returned with the request.
+     *   Laving the default value of the parameter to 0 will result in not imposing any
+     *   explicit restrictions on a size of the result set. NOte that other, resource-defined
+     *   restrictions will still apply. The later includes the maximum size of the Google Protobuf
+     *   objects, the amount of available memory, etc.
+
+     * @param priority
+     *   priority level of the request
+     *
+     * @param keepTracking
+     *   keep tracking the request before it finishes or fails
+     *
+     * @param messenger
+     *   interface for communicating with workers
      */
-    SqlResultSet const& responseData() const;
+    SqlBaseRequest(ServiceProvider::Ptr const& serviceProvider,
+                   boost::asio::io_service& io_service,
+                   std::string const& worker,
+                   uint64_t maxRows,
+                   int priority,
+                   bool keepTracking,
+                   std::shared_ptr<Messenger> const& messenger);
+
+    /// @see Request::startImpl()
+    void startImpl(util::Lock const& lock) final;
+
+    /// @see Request::savePersistentState()
+    void savePersistentState(util::Lock const& lock) final;
+
+    /**
+     * Request body to be sent to the worker. The content of the request is partially
+     * set by this class's constructor, and it's fully initialized by the constructor
+     * of a subclass depending on a type of request.
+     */
+    ProtocolRequestSql requestBody;
+
+private:
+
+    /**
+     * Start the timer before attempting the previously failed
+     * or successful (if a status check is needed) step.
+     *
+     * @param lock
+     *   a lock on Request::_mtx must be acquired before calling this method
+     */
+    void _wait(util::Lock const& lock);
+
+    /**
+     * Callback handler for the asynchronous operation
+     *
+     * @param ec
+     *   error code to be checked
+     */
+    void _awaken(boost::system::error_code const& ec);
```

Comment:
I think `_handleAsync` is too generic because virtually all operations performed in a context of the request classes are asynchronous by the very nature of the Replication Framework (based on `BOOST ASIO`). Also, after I rename method `_wait` into `_waitAsync` the meaning of method `_awaken` becomes a bit more apparent. I'm also going to reject it more clear in the method documentation.